accepted/file-enumeration.md

```diff
@@ -0,0 +1,238 @@
+    /// </summary>
+    public ref struct RawFindData
+    {
+        // This will have private members that hold the native data and
+        // will lazily fill in data for properties where such data is not
+        // immediately available in the current platform's native results.
+
+		// The full path to the directory the current result is in
+        public string Directory { get; }
+
+        // The full path to the starting directory for enumeration
+        public string OriginalDirectory { get; }
+
+        // The path to the starting directory as passed to the enumerable constructor
+        public string OriginalUserDirectory { get; }
+
+        // Note: using a span allows us to reduce unneeded allocations
+        public ReadOnlySpan<char> FileName { get; }
+        public FileAttributes Attributes { get; }
+        public long Length { get; }
+
+        public DateTime CreationTimeUtc { get; }
+        public DateTime LastAccessTimeUtc { get; }
+        public DateTime LastWriteTimeUtc { get; }
+    }
+}
+```
+
+### Transforms & Predicates
+
+We'll provide common predicates transforms for building searches.
+
+``` C#
+namespace System.IO
+{
+    internal static partial class FindPredicates
+    {
+        internal static bool NotDotOrDotDot(ref RawFindData findData)
+        internal static bool IsDirectory(ref RawFindData findData)
+    }
+
+    public static partial class FindTransforms
+    {
+        public static DirectoryInfo AsDirectoryInfo(ref RawFindData findData)
+        public static FileInfo AsFileInfo(ref RawFindData findData)
+        public static FileSystemInfo AsFileSystemInfo(ref RawFindData findData)
+        public static string AsFileName(ref RawFindData findData)
+        public static string AsFullPath(ref RawFindData findData)
+    }
+}
+
+```
+
+### DosMatcher
+
+We currently have an implementation of the algorithm used for matching files on Windows in FileSystemWatcher. Providing this publicly will allow consistently matching names cross platform according to Windows rules if such behavior is desired.
+
+``` C#
+namespace System.IO
+{
+    public static class DosMatcher
+    {
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        public unsafe static string TranslateExpression(string expression)
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+    }
+}
+```
+
+### Samples
+
+Getting full path of all files matching a given name pattern (close to what FindFiles does, but returning the full path):
+
+``` C#
+public static FindEnumerable<string, string> GetFiles(string directory,
```

Comment:
This sample has a heavy cognitive load:

- Tuples
- ref
- Two delegates, likely wrapping other BCL API, which intellisense won't help find
- concept of passing in a state at the start of an enumeration

It's the ultimate API, but it is likely intimidating for someone who simply wants look for files matching a regex, without caring about every drop of performance

Reply:
For sure. Doing this doesn't preclude us adding simpler overloads, and I would actually expect to eventually.

---
In reply to: [154808558](https://github.com/dotnet/designs/pull/24#discussion_r154808558) [](ancestors = 154808558)

Reply:
I think it would be helpful to propose those here.